### ISSUE TITLE
Integrate into Home Assistant core

### ISSUE THREAD

**Issue by mammuth** [mammuth/ha-fritzbox-tools#29]
On the long run it'd be cool to integrate this component into HA itself.

This ticket simply serves as a brainstorming area what should be done in order to integrate it

- [x] Async (https://github.com/mammuth/ha-fritzbox-tools/pull/27)
- [x] Config validation (https://github.com/mammuth/ha-fritzbox-tools/pull/28)
- [x] Support for config flow (https://github.com/mammuth/ha-fritzbox-tools/issues/30)
- [x] Format with black
- [ ] integration of some other fritz components (currently residing in core): `fritz`, `fritzbox`, `fritzbox_callmonitor`, `fritzbox_netmonitor`, `fritzdect`
- [ ] Tests (https://github.com/mammuth/ha-fritzbox-tools/issues/31)
- [x] Wait some more time so more people can test the functionality in this component
- [ ] ... ideas welcome! ...

**Comment by AaronDavidSchneider** [comment]
> * [ ]  integration of some other fritz components (currently residing in core): `fritz`, `fritzbox`, `fritzbox_callmonitor`, `fritzbox_netmonitor`, `fritzdect`

do you want to combine the components into one large component? I think this is something that should be done after an initial release of the component in the core.
Such a huge pull request won't be merged otherwise.

**Comment by mammuth** [comment]
I don't know. But we should think about it. I'm with you in regards of the dect and monitor stuff, but we might integrate the device tracker from beginning for example. 

We could also ask more senior contributors on their opinion.

**Comment by AaronDavidSchneider** [comment]
I would totally agree with you! However this would conflict the developer documents:
https://developers.home-assistant.io/docs/en/creating_component_code_review.html#5-make-your-pull-request-as-small-as-possible

**Comment by mammuth** [comment]
I think we could start preparing the integration into core.

Some questions/things to think about:
- Should we create a new component next to `fritz`, `fritzbox`, `fritz_dect`, etc. or override one of them? We should probably ask a core contributor what's preferred.
- Which platforms/features do we include in the initial PR (to keep the diff as small as possible)
- ...?

**Comment by da-anda** [comment]
I'm so glad I found this, after searching all over the web on how to add a guest wifi toggle to HA. I would very welcome if this would become a core component. IMO the way to go would be to merge all the currently cluttered components into one namespace on the long run, but start small for now and try to get feedback from core devs ASAP.
Adding yet another component namespace would IMO be the wrong approach, so it's probably best to override a current one. There I'd probably use `fritz` since this is the most suiteable one for a platform which supports a multitude of devices (FritzBox, FritzDect, ...) all sharing a common API (FritzOS). When overriding a namespace, the new implementation ofc would have to support the existing features of the namespace it's overriding, or chances if it to be accepted into the core are probably quite low.

**Comment by mammuth** [comment]
There is a new PR which adds config flows to core: https://github.com/home-assistant/home-assistant/pull/31240

**Comment by escoand** [comment]
My far goal was also to merge all the different integrations. imo `fritzdect` could be dropped immediately.

**Comment by jankatins** [comment]
One more thing would IMO to enable multiple fritzboxes in one network (https://github.com/mammuth/ha-fritzbox-tools/issues/72) as it might require some user visible breaking config changes.

**Comment by AaronDavidSchneider** [comment]
What do you think @mammuth, when would be the time to merge into HA core?
Are there issues left we should address first? 
Do you want to open up a PR? I could do too, if you want me to.

**Comment by mammuth** [comment]
I think there are no issues left. But since this integration became quite large already we need to find a good path to move functionality over to core in small pieces. 

What do we start with?

**Comment by AaronDavidSchneider** [comment]
That sounds good!

I think the best way would be to setup a few PRs. 
The first could be the WiFi switch (including all of the config flow etc)

Then dependent on this PR, we could add all other features into single PRs?

The disadvantage is that we would need to rewrite parts of the config flow to match the individual PRs...

Or we try one large PR and see what happens 😂

**Comment by mammuth** [comment]
I think one large PR is just being closed :wink:

**Comment by AaronDavidSchneider** [comment]
Probably true... 😂

**Comment by chemelli74** [comment]
Guys, I spoke with Paulus and the answer is:

> Small prs and added to existing Frits integration


Thank you for your great work!

Simone

**Comment by AaronDavidSchneider** [comment]
> Guys, I spoke with Paulus and the answer is:
> 
> > Small prs and added to existing Frits integration
> 
> Thank you for your great work!
> 
> Simone

Puh... That sounds like a lot of work..

**Comment by chemelli74** [comment]
I try to resurrect this old thread ;-)

What if we add one feature at a time to the current core implementation ?
Being part of Shelly integration Team, I learned it takes a while to get used to all coding rules and I strongly think this is the better way to go.

I may help if you think you need a contributor.

Simone

**Comment by mammuth** [comment]
I guess the current first challenge is to understand how to get started here. Especially since the the core integration might be lacking some basics like a more sophisticated config flow, base classes like we have it etc. 

Besides, everyone can obviously pick any feature that we found as part of this custom integration and port it into core. So if there are no blockers for you, go for it.

**Comment by chemelli74** [comment]
> Besides, everyone can obviously pick any feature that we found as part of this custom integration and port it into core. So if there are no other blockers for you, go for it.

I was more hoping in a Team job ;-)

**Comment by mammuth** [comment]
Sure. I wanted to emphasize that everyone can do it, without being blocked by me or other developers of this integration. 

Also, disclaimer: I'm currently lacking motivation and time to kickoff this larger project on my own.

**Comment by AaronDavidSchneider** [comment]
I can only agree. I think my supervisors would prefer if I focus on my actual work ...

**Comment by chemelli74** [comment]
I was thinking we could start with `config_flow.py` and `binary_sensor.py`; quite small but a good base for the next actions.

**Comment by chemelli74** [comment]
WiP:

```
root ➜ /workspaces/core/homeassistant/components/fritz (chemelli74-newfritzfeatures) $ pre-commit run
pyupgrade................................................................Passed
black....................................................................Passed
codespell................................................................Passed
flake8...................................................................Passed
bandit...................................................................Passed
isort....................................................................Passed
Check JSON...........................................(no files to check)Skipped
Don't commit to branch...................................................Passed
yamllint.............................................(no files to check)Skipped
mypy.....................................................................Passed
gen_requirements_all.................................(no files to check)Skipped
hassfest.............................................(no files to check)Skipped
```

**Comment by chemelli74** [comment]
Can I list you 2 as codeowners? Should I add myself ?

Simone

**Comment by chemelli74** [comment]
`config_flow` and `binary sensor` should be now ready.

Missing  `device_tracker` integration in config_flow and then I'll be ready for first PR.
This will take a bit, as I miss the basis of device_tracker, unfortunately.

Simone

**Comment by chemelli74** [comment]
Slowly progressing:

![image](https://user-images.githubusercontent.com/57354320/111069900-8e929580-84cf-11eb-9de3-6e3a0d5cbb78.png)

Simone

**Comment by chemelli74** [comment]
Submitted a few PR to align code to HASS pre-commit.
If you agree and commit them, then my plan is:

- create a router.py where to move classes like `FritxboxTools` ( I have the code here and working)
- create a draft PR for `device_tracker` so to get help.

For device tracker I'm stuck on:

- Device I/O in event loop
- Device tracker polling update

Simone

**Comment by chemelli74** [comment]
Cleanup is fully merged and even released. Thx a lot to @mammuth ;-)

My current job for `device_tracker` is basically working: [chemelli74-device_tracker branch](https://github.com/mammuth/ha-fritzbox-tools/compare/master...chemelli74:chemelli74-device_tracker?expand=1)

If you have wish, would be nice to get help on :

- Device I/O in event loop -> common.py line 214
- Device tracker polling update

Simone

**Comment by chemelli74** [comment]
Today commit fixed the first issue: I/O in event loop 🥳 

Now only 1 thing missing:

- Device tracker polling update

Simone

**Comment by chemelli74** [comment]
Code completed and cleaned up!

Let's share together how to proceed.
@mammuth, @AaronDavidSchneider,  you find me on discord. Maybe a group there would help.

Simone

**Comment by TheZoker** [comment]
Can confirm, works great!

**Comment by escoand** [comment]
Thanks, also working for me, but 2 hints:

- The default host should be `fritz.box` or `169.254.1.1`. This should be valid for all standard installations.
- The default username should be `admin`. For a normal user it's not obvious what's the username because you don't need one on the web interface.

**Comment by jankatins** [comment]
FYI, the username becomes more important with the newer 7.25 version: https://avm.de/service/schnittstellen/

> If the login type "Login only with FRITZ!Box password" was used up until now, a FRITZ! user
will be created automatically during the update. This FRITZ! user has a randomly chosen
name based on the scheme fritz[0-9] {4}, for example fritz1234, fritz1337 or fritz0042.

**Comment by mammuth** [comment]
> The default username should be admin. For a normal user it's not obvious what's the username because you don't need one on the web interface.

If I remember it correctly, we got some reports in the past that functionality didn't work for password-only user accounts. That's why we just went with the suggestion to create a dedicated user. 
Not sure whether this is still required, though...

> Let's share together how to proceed.
@mammuth, @AaronDavidSchneider, you find me on discord. Maybe a group there would help.

Good idea: https://discord.gg/WAxZKTfTc9
